File: 0496-Next Greater Element I/0496-Next Greater Element I.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& findNums, vector<int>& nums) {
        int m = findNums.size();
        if (m == 0)
        {
            return vector<int>();
        }
        
        vector<int> result(m);
        unordered_map<int, int> table;
        stack<int> St;
        for (int num : nums)
        {
            while (!St.empty() && St.top() < num)
            {
                table[St.top()] = num;
                St.pop();
            }
            St.push(num);
        }
        
        for (int i = 0; i < m; ++i)
        {
            if (table.find(findNums[i]) != table.end())
            {
                result[i] = table[findNums[i]];
            }
            else
            {
                result[i] = -1;
            }
        }
        
        return result;
    }
};

File: 0401-0500/0496-Next Greater Element I/0496-Next Greater Element I.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        St = []
        table = {}
        for num in nums2:
            while St and St[-1] < num:
                table[St.pop()] = num
            St.append(num)
            
        return [table[num] if num in table else -1 for num in nums1]
